worlds/cvcotm/__init__.py

```diff
@@ -0,0 +1,228 @@
+    }
+    location_name_groups = get_location_name_groups()
+    options_dataclass = CVCotMOptions
+    options: CVCotMOptions
+    settings: typing.ClassVar[CVCotMSettings]
+    hint_blacklist = {lname.ba24}  # The Battle Arena reward, if it's put in, will always be a Last Key.
+
+    item_name_to_id = {name: cvcotm_item_info[name].code + BASE_ID for name in cvcotm_item_info
+                       if cvcotm_item_info[name].code is not None}
+    location_name_to_id = get_location_names_to_ids()
+
+    # Default values to possibly be updated in generate_early
+    total_last_keys: int = 0
+    required_last_keys: int = 0
+
+    auth: bytearray
+
+    web = CVCotMWeb()
+
+    def generate_early(self) -> None:
+        # Generate the player's unique authentication
+        self.auth = bytearray(self.multiworld.random.getrandbits(8) for _ in range(16))
+
+        # If playing with the Completion Goal option set to Battle Arena while the Required Skirmishes option is set to
+        # All Bosses And Arena, set the latter down to just All Bosses instead. The Location in the Ceremonial Room
+        # would require the Goal to be sent before it can be accessed at all, and that would be weird.
+        if self.options.completion_goal == CompletionGoal.option_battle_arena and self.options.required_skirmishes == \
+                RequiredSkirmishes.option_all_bosses_and_arena:
+            logging.warning(f"[{self.multiworld.player_name[self.player]}] Requiring a Last Key to be obtained from "
+                            f"the Battle Arena while requiring just the Battle Arena to be completed for Goal "
+                            f"completion makes the Ceremonial Room item impossible to obtain without completing the "
+                            f"Goal first. Changing Required Skirmishes to just All Bosses.")
+            self.options.required_skirmishes = RequiredSkirmishes.option_all_bosses
+
+        # If Required Skirmishes are on, force the Required and Available Last Keys to 8 or 9 depending on which option
+        # was chosen.
+        if self.options.required_skirmishes == RequiredSkirmishes.option_all_bosses:
+            self.options.required_last_keys.value = 8
+            self.options.available_last_keys.value = 8
+        elif self.options.required_skirmishes == RequiredSkirmishes.option_all_bosses_and_arena:
+            self.options.required_last_keys.value = 9
+            self.options.available_last_keys.value = 9
+
+        self.total_last_keys = self.options.available_last_keys.value
+        self.required_last_keys = self.options.required_last_keys.value
+
+        # If there are more Last Keys required than there are Last Keys in total, drop the required Last Keys to
+        # the total Last Keys.
+        if self.required_last_keys > self.total_last_keys:
+            self.required_last_keys = self.total_last_keys
+            logging.warning(f"[{self.multiworld.player_name[self.player]}] The Required Last Keys "
```

Comment:
Again I'll just check that this behavior is from what players seem to want? I ask here because this is the opposite of how I would expect such options to resolve (if they resolved at all instead of throwing a full error)

Reply:
I think my thought process here went something like "oh, if players made there be more required keys than there are in total, then I don't want players to feel intimidated that they have to find more keys than what they put into the pool". In my experience, the more instances of something needed to finish when there are 0 redundant extras of that thing in the pool, the later I am to finish because there's a higher chance of one of those things landing somewhere evil.